cluster.c

```diff
@@ -379,6 +379,58 @@ RRStatus ShardGroupAppendLogEntry(RedisRaftCtx *rr, ShardGroup *sg, int type, vo
     return RR_OK;
 }
 
+RRStatus ShardGroupsAppendLogEntry(RedisRaftCtx *rr, int num_sg, ShardGroup **sg, int type, void *user_data)
+{
+    /* Make sure we're still a leader, could have changed... */
+    /* SJP: unsure this is true in this case */
+    if (!raft_is_leader(rr->raft)) {
+        return RR_ERROR;
+    }
+
+    /* Serialize */
+    /* We reuse the existing shard group serialization
+     * Format:
+     * <num shard groups>:strlen(sg #1 payload):sg #1 pauload|...:strlen(sg #n payload):sg #n payload|
+     */
+    RedisModuleString * str = RedisModule_CreateStringPrintf(rr->ctx, "%d:", num_sg);
+    for (int i = 0; i < num_sg; i++) {
+        char *payload = ShardGroupSerialize(sg[i]);
+        RedisModuleString *payload_len = RedisModule_CreateStringPrintf(rr->ctx, "%lu:", strlen(payload));
+        size_t payload_len_str_len;
+        const char * payload_len_str = RedisModule_StringPtrLen(payload_len, &payload_len_str_len);
+        RedisModule_StringAppendBuffer(rr->ctx, str, payload_len_str, payload_len_str_len);
+        RedisModule_StringAppendBuffer(rr->ctx, str, payload, strlen(payload));
+        RedisModule_StringAppendBuffer(rr->ctx, str, ":", 1);
+        if (!payload) {
+            RedisModule_FreeString(rr->ctx, str);
+            return RR_ERROR;
+        }
+        RedisModule_Free(payload);
+    }
+
+    size_t len;
+    const char * payload = RedisModule_StringPtrLen(str, &len);
+
+    raft_entry_t *entry = raft_entry_new(len);
+    entry->type = type;
+    entry->id = rand();
+    entry->user_data = user_data;
+    memcpy(entry->data, payload, len);
```

Comment:
Are we copying here without NULL character? Should we copy it as well if there is no downside?

Reply:
well, this is an artifact before of using RedisModule Strings which aren't guaranteed to be null terminated, with your suggestion above, that changes.

Reply:
thinking about it, I'm unsure in practice it gains us anything even with the change above, but am open to being convinced I'm wrong

Reply:
I think its fine as is.  in general entry data can't be assumed to be a C string, so useful to always depend on data_len part of it, and not sometimes C string and sometimes not as then it gets confusing.